src/ML-POD/fastpod.cpp

```diff
@@ -0,0 +1,4249 @@
+            cU[i1 + nelements*(n1+q) + nelements*K3*m] += pc3[n1+q]*c2*c3;
+            k += 1;
+          }
+        }
+      }
+    }
+  return e;
+}
+
+void FASTPOD::twobodydescderiv(double *d2, double *dd2, double *rbf, double *rbfx,
+        double *rbfy, double *rbfz, int *tj, int N)
+{
+  for (int m=0; m<nl2; m++)
+    d2[m] = 0.0;
+  for (int m=0; m<3*N*nl2; m++)
+    dd2[m] = 0.0;
+
+  for (int m=0; m<nrbf2; m++) {
+    for (int n=0; n<N; n++) {
+      int i2 = n + N*m;
+      int i1 = n + N*m + N*nrbf2*(tj[n]-1);
+      d2[m + nrbf2*(tj[n]-1)] += rbf[i2];
+      dd2[0 + 3*i1] += rbfx[i2];
+      dd2[1 + 3*i1] += rbfy[i2];
+      dd2[2 + 3*i1] += rbfz[i2];
+    }
+  }
+}
+
+void FASTPOD::twobodydescriptors(double *d2, double *dd2, double *rij, double *temp, int *tj, int Nj)
+{
+  orthogonalradialbasis(temp, rij, Phi, besselparams, rin, rcut-rin,
+          pdegree[0], pdegree[1], nbesselpars, nrbf2, Nj);
+
+  int n2 = Nj*nrbf2;
+  twobodydescderiv(d2, dd2, temp, &temp[n2], &temp[2*n2], &temp[3*n2], tj, Nj);
+}
+
+void FASTPOD::twobodydescriptors(double *gd2, double *gdd2, double *d2, double *dd2, double *rij,
+        double *temp, int *ai, int *aj, int *ti, int *tj, int Nj, int natom)
+{
+  orthogonalradialbasis(temp, rij, Phi, besselparams, rin, rcut-rin,
+          pdegree[0], pdegree[1], nbesselpars, nrbf2, Nj);
+
+  int n2 = Nj*nrbf2;
+  twobodydescderiv(d2, dd2, temp, &temp[n2], &temp[2*n2], &temp[3*n2], tj, Nj);
+
+  tallytwobodyglobdesc(gd2, d2, elemindex, nrbf2, nelements, ti[0]-1);
+
+  tallytwobodyglobdescderiv(gdd2, dd2, ai, aj, ti, tj, elemindex, nrbf2, nelements, natom, Nj);
+}
+
+
+void FASTPOD::onebodydescriptors(double *gd1, double *gdd1, int *ti, int natom, int i)
+{
+  for (int m=0; m<nelements; m++) {
+    double onehot = (ti[0] == (m+1)) ? 1.0 : 0.0;
+    gd1[m] += onehot;
+    gdd1[0 + 3*(i + natom*m)] = 0.0;
+    gdd1[1 + 3*(i + natom*m)] = 0.0;
+    gdd1[2 + 3*(i + natom*m)] = 0.0;
+  }
+}
+
+double FASTPOD::onebodyenergy(double *coeff1, int *ti)
+{
+  double e1 = 0.0;
+  for (int m=0; m<nelements; m++)
+    e1 += (ti[0] == (m+1)) ? coeff1[m] : 0.0;
+
+  return e1;
+}
+
+void FASTPOD::radialfunctions(double *rbf, double *rij, double *besselparams, double rin,
+        double rmax, int besseldegree, int inversedegree, int nbesselpars, int N)
+{
+  for (int n=0; n<N; n++) {
+    double rij1 = rij[0+3*n];
+    double rij2 = rij[1+3*n];
+    double rij3 = rij[2+3*n];
+
+    double dij = pow(rij1*rij1 + rij2*rij2 + rij3*rij3, 0.5);
+    double r = dij - rin;
+    double y = r/rmax;
+    double y2 = y*y;
+    double y3 = 1.0 - y2*y;
+    double y4 = y3*y3 + 1e-6;
+    double y5 = pow(y4, 0.5);
+    double y6 = exp(-1.0/y5);
+    double fcut = y6/exp(-1.0);
+
+    for (int j=0; j<nbesselpars; j++) {
+      double x =  (1.0 - exp(-besselparams[j]*r/rmax))/(1.0-exp(-besselparams[j]));
+      for (int i=0; i<besseldegree; i++)
+        rbf[n + N*i + N*besseldegree*j] = ((sqrt(2.0/(rmax))/(i+1)))*fcut*sin((i+1)*M_PI*x)/r;
```

Comment:
```suggestion
        rbf[n + N*i + N*besseldegree*j] = ((sqrt(2.0/(rmax))/(i+1)))*fcut*sin((i+1)*MY_PI*x)/r;
```

Using `M_PI` is not portable and may introduce platform specific inconsistencies. You should use `MY_PI` instead like elsewhere.